zuko/transforms.py

```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
+        self.dbasis = self.get_bernstein_basis(
+            self.order - 1, device=theta.device, dtype=theta.dtype
+        )
+
+        # save slope on boundaries for interpolation
+        x = torch.tensor([self.eps, 1 - self.eps], device=theta.device, dtype=theta.dtype)
+        rank = self.theta.dim()
+        if rank > 1:
+            # add singleton batch dimensions
+            dims = [...] + [None] * (rank - 1)
+            x = x[dims]
+        self.offset = self.b_poly(x, self.theta, self.basis)
+        self.slope = self.b_poly(x, self.dtheta, self.dbasis)
 
     @staticmethod
-    def _increasing(theta: Tensor) -> Tensor:
+    def _increasing(unconstrained_theta: Tensor, smooth_bounds: bool) -> Tensor:
         r"""Processes the unconstrained output of the hyper-network to be increasing."""
 
-        shift = math.log(2.0) * theta.shape[-1] / 2
+        shift = math.log(2.0) * unconstrained_theta.shape[-1] / 2
+
+        theta_min = unconstrained_theta[..., :1]
+        unconstrained_theta = unconstrained_theta[..., 1:]
+
+        if smooth_bounds:
+            unconstrained_theta = torch.cat((
+                unconstrained_theta[..., :1],
+                unconstrained_theta,
+                unconstrained_theta[..., -1:],
+            ))
+        widths = torch.nn.functional.softplus(unconstrained_theta)
+
+        widths = torch.cat((theta_min, widths), dim=-1)
+
+        theta = torch.cumsum(widths, dim=-1) - shift
+
+        return theta
 
-        widths = torch.nn.functional.softplus(theta[..., 1:])
-        widths = torch.cat((theta[..., :1], widths), dim=-1)
+    @staticmethod
+    def get_bernstein_basis(order, **kwds):
+        alpha = torch.arange(1, order + 2, **kwds)
+        beta = torch.arange(order + 1, 0, -1, **kwds)
+        basis = torch.distributions.Beta(alpha, beta)
+        return basis
+
+    @staticmethod
+    def b_poly(x, theta, basis):
+        b = basis.log_prob(x.unsqueeze(-1)).exp()
+        y_poly = torch.mean(b * theta, dim=-1)
+        return y_poly
+
+    def scale_data(self, x):
+        x = (x + self.bound) / (2 * self.bound)  # map [-B, B] to [0, 1]
+
+        return x
 
-        return torch.cumsum(widths, dim=-1) - shift
+    def unscale_data(self, y):
+        y = y * 2 * self.bound - self.bound  # map [0, 1] to [-B, B]
+
+        return y
+
+    def _extrapolation(self, x: Tensor):
+        y0 = self.slope[0] * (x - self.eps) + self.offset[0]  # h'(eps) * x + h(eps)
+        y1 = self.slope[1] * (x - 1 + self.eps) + self.offset[1]  # h'(1-eps) * x + h(1-eps)
+        return y0, y1
 
     def f(self, x: Tensor) -> Tensor:
-        if self.linear:
-            x = (x + self.bound) / (2 * self.bound)  # map [-B, B] to [0, 1]
-        else:
-            x = torch.nn.functional.sigmoid(x)  # map [-inf, inf] to [0, 1]
+        x = self.scale_data(x)
+
+        left_bound = x <= self.eps
+        right_bound = x >= 1 - self.eps
+        x_safe = torch.where(left_bound | right_bound, 0.5, x)
```

Comment:
I am afraid that the `0.5` might not work with all supported versions of PyTorch. To be tested. Also, `left_bound` and `right_bound` are not great names as they are out-of-bound indicators.

Reply:
Would this be better:

            x_safe = torch.where(lower_bound | upper_bound, 0.5 * torch.ones_like(x), x)